docs/best-practices/index.md

```diff
@@ -0,0 +1,438 @@
+
+For more information about state management when using Terraform, refer to [Hashicorp's online documentation](https://developer.hashicorp.com/terraform/language/state).
+
+### Don't Modify State Directly
+
+Directly modifying Terraform state files is strongly discouraged due to the critical role they play in Terraform's management of infrastructure resources.
+
+The state file acts as a single source of truth for both the configuration and the real-world resources it manages, which is critical when Terraform calculates the differences between "intended" and "actual" configuration when running `terraform plan`.
+
+Manual edits can lead to inconsistencies between the actual state of your infrastructure and Terraform's record, potentially causing unresolvable conflicts in the plan phase, or may result in errors when the provider is reading/writing the state of a resource. Such actions undermine the integrity of your infrastructure management, leading to difficult-to-diagnose issues, resource drift, and potentially the loss or corruption of critical infrastructure.
+
+Instead of directly editing state files, it is best practice to use Terraform's built-in commands such as `terraform state rm` or `terraform import` to safely make changes. This approach ensures that Terraform can accurately track and manage the state of provisioned infrastructure, maintaining the reliability and predictability of your infrastructure as code environment.
+
+For more information about state management when using Terraform, refer to [Hashicorp's online documentation](https://developer.hashicorp.com/terraform/language/state).
+
+### Ensure Provider Warnings are Captured and Reviewed
+
+Terraform providers can produce warnings as a result of operations such as `terraform validate`, `terraform plan` and `terraform apply`.  When these operations are run using the CLI, the warnings are directed to the command line output and when these operations are run in cloud services such as Terraform Cloud, these warnings are shown in the UI.
+
+Ping Identity's Terraform providers can show warnings that need to be captured and reviewed.  For example, the PingOne Terraform provider will produce warnings when specific configuration is used that remove guardrails to prevent accidental deletion of data.
+
+It is highly recommended that warnings shown on the `terraform plan` stage especially are captured reviewed before the `terraform apply` stage is run, as the messages could be alerting the administrator to potential undesired results of the `terraform apply` stage.
+
+## HCL Writing Recommendations
+
+### Use Terraform Formatting Tools
+
+When writing Terraform HCL, using `terraform fmt` is a straightforward yet powerful practice.  `terraform fmt` and equivalent formatting tools adjusts the Terraform code to a standard style, which helps keep the codebase tidy and consistent.  Typically, this means maintaining consistent indentation, spacing and alignment of code.
```

Comment:
tools **adjust** the Terraform

style, **keeping the codebase consistent.**. Typically, this **formatting deals with** maintaining ...